src/core/net/nat64_translator.cpp

```diff
@@ -226,14 +240,44 @@ Translator::Result Translator::TranslateToIp6(Message &aMessage)
         ExitNow(res = kDrop);
     }
 
-    mapping = FindMapping(ip4Header.GetDestination());
+    switch (ip4Header.GetProtocol())
+    {
```

Comment:
It may help if we introduces a `Ip4::Headers` type (which contains the IPv4 header along with union for TCP/UDP and ICMP headers) similar to `Ip6::Headers`. 

I feel may make the code simpler? We can pass `Ip4::Headers` to `FindOrAllocateMatching()` method to find a match based on all parsed headers for us?

Reply:
Yes, that will be a good improvement. Same question as above, should I create a separate PR for this?

Reply:
Submitted #11151 to address this improvement.